### PR TITLE
add toast to transactions

### PR DESCRIPTION
**Pull request by vanillaHill**
Closes https://github.com/FortressFinance/issues/issues/220

### CODE REVIEW

File: src/components/Layout.tsx [hunk 1]
```diff
@@ -43,20 +44,14 @@ const Layout: FC<PropsWithChildren> = ({ children }) => {
...
-        <Toaster
-          position="top-right"
-          containerClassName="mr-[2.5%] -mt-[5px]"
-        />
-      </div>
-      <div className="md:hidden">
-        {/* mobile toaster */}
-        <Toaster position="bottom-center" />
-      </div>
+      <Toaster
```
Comment (vanillaHill) [inline src/components/Layout.tsx:51]: The previous method using the className introduces a bug (toast position overlaps)

File: src/components/VaultRow/lib/VaultDepositForm.tsx [hunk 2]
```diff
@@ -141,17 +185,21 @@ export const VaultDepositForm: FC<VaultProps> = (props) => {
...
-    onSuccess: onDepositSuccess,
+    hash: depositUnderlyingTxHash,
+    onSuccess: () => onDepositSuccess(depositUnderlyingTxHash),
+    onError: () => onDepositError(depositUnderlyingTxHash),
   })
 
   // Form submit handler
   const onSubmitForm: SubmitHandler<TokenFormValues> = async () => {
     if (requiresApproval) {
+      onApproveStart()
```
Comment (vanillaHill) [inline src/components/VaultRow/lib/VaultDepositForm.tsx:198]: Can't use onSettled method, onSuccess is fired before it. I workaround it by calling the loading here.

File: src/components/Toast/Toast.tsx [hunk 3]
```diff
@@ -32,7 +32,7 @@ export const Toast: FC<PropsWithChildren<ToastComponentProps>> = ({
     >
       <div
         className={clsxm(
-          "w-full max-w-xs rounded-md bg-gradient-radial from-[#61312A] to-[#5D2741] p-4 text-white shadow",
+          "w-auto rounded-md bg-gradient-radial from-[#61312A] to-[#5D2741] p-4 text-white shadow",
```
Comment (aquaticone) [inline src/components/Toast/Toast.tsx:35]: Is `w-auto` doing anything here?

File: src/components/VaultRow/lib/VaultDepositForm.tsx [hunk 4]
```diff
@@ -32,6 +35,15 @@ export const VaultDepositForm: FC<VaultProps> = (props) => {
   const { address: userAddress } = useAccount()
   const chainId = useActiveChainId()
   const vault = useVault(props)
+  const toastManager = useToast()
+  const [depositToastId, setDepositToastId] = useState<string | undefined>()
+  const [approveToastId, setApproveToastId] = useState<string | undefined>()
+  const depositLoadingMsg = "Waiting for deposit transaction..."
```
Comment (aquaticone) [inline src/components/VaultRow/lib/VaultDepositForm.tsx:41]: No need to have variables for these -- just define the message inline when invoking the toasts

File: src/components/VaultRow/lib/VaultDepositForm.tsx (hunk 2 above)
Comment (aquaticone) [reply src/components/VaultRow/lib/VaultDepositForm.tsx:198]: There is a cleaner way to handle this without `useState` by invoking `writeAsync`:

```ts
const waitingForSigner = toastManager.loading("Waiting for signature")
await approve
  .writeAsync?.()
  .then((receipt) =>
    // this fires after the transaction has been broadcast successfully
    toastManager.loading(
      "Waiting for transaction confirmation",
      receipt.hash
    )
  )
  .catch((err) =>
    // this fires after a failure to broadcast the transaction
    toastManager.error(
      err instanceof UserRejectedRequestError
        ? "User rejected request"
        : "Error broadcasting transaction"
    )
  )
  .finally(() => toastManager.dismiss(waitingForSigner))
```

In the case of "waiting for signature" toasts, we don't have a hash so they must be manually dismissed -- but for toasts we *do* have a hash for, I really think you should try self-dismissal within the `LoadingToast` component. 

It is 100% worth implementing it this way as to remove any need to manually keep track of toast IDs and grealy streamline the entire process for us while implementing.

You'll need to accept the toast `id` and `txHash` as props on `LoadingToast` and wait for the transaction there:

```ts
// this is all you need to do; hook will not run if `hash` is undefined
useWaitForTransaction({ hash, onSettled: () => toast.dismiss(id) })
```

File: src/components/VaultRow/lib/VaultDepositForm.tsx [hunk 5]
```diff
@@ -90,8 +129,11 @@ export const VaultDepositForm: FC<VaultProps> = (props) => {
     enabled: requiresApproval,
   })
   const approve = useContractWrite(prepareApprove.config)
+  const approveTxHash = approve.data?.hash
   const waitApprove = useWaitForTransaction({
-    hash: approve.data?.hash,
+    hash: approveTxHash,
+    onSuccess: () => onApproveSuccess(approveTxHash),
```
Comment (aquaticone) [inline src/components/VaultRow/lib/VaultDepositForm.tsx:135]: You can delete the variables you're using to store hashes and simply do this:

```ts
const waitApprove = useWaitForTransaction({
  hash: approve.data?.hash,
  onSettled: (receipt, error) =>
    error
      ? toastManager.error(
          "Transaction unsuccessful",
          receipt?.transactionHash
        )
      : toastManager.success(
          "Transaction successful",
          receipt?.transactionHash
        ),
})
```

File: src/hooks/useMedia.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import { useMediaQuery } from "react-responsive"
+
+export function useMedia(query: string) {
```
Comment (aquaticone) [inline src/hooks/useMedia.ts:3]: Why don't you just use their hook without this wrapper?

**Review by aquaticone — CHANGES_REQUESTED** [review]

File: src/components/Toast/Toast.tsx [hunk 7]
```diff
@@ -11,7 +12,7 @@ export type ToastComponentProps = {
 
 export type TransactionToastComponentProps = ToastComponentProps & {
   message: string
-  txHash?: string
+  txHash?: Address
```
Comment (aquaticone) [inline src/components/Toast/Toast.tsx:15]: I think probably just make this accept a string so you don't need to tell TS `txHash as Address` everywhere

**Review by aquaticone — APPROVED** [review]
Looks great, works well. Good stuff.

Just one last little nitpick but you're good to go.

**Pull request merged**